Random crash at startup

I'm suddenly getting crashes after starting the game and then choosing multiplayer.
The snippet is from release-20141029 which crashes immediately when I choose multiplayer.
The exception is identical with playtest-20150405 and playtest-20150408 when I choose "incompatible" in multiplayer mode. It does not crash the playtests if I do not choose to see the "incompatible" servers.

<pre><code>
Platform is Linux
Using SDL 2 with OpenGL renderer
Desktop resolution: 1920x1200
No custom resolution provided, using desktop resolution
Using resolution: 1920x1200
Detected OpenGL version: 3.0
Using OpenAL sound engine
Using default device
Available mods:
        cnc: Tiberian Dawn (release-20141029)
        d2k: Dune 2000 (release-20141029)
        modchooser: Mod Chooser (release-20141029)
        ra: Red Alert (release-20141029)
Loading mod: ra
Exception of type `System.ArgumentException`: A null reference or invalid value was found [GDI+ status: InvalidParameter]
  at System.Drawing.GDIPlus.CheckStatus (Status status) [0x00000] in <filename unknown>:0 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format, System.Drawing.Imaging.BitmapData bitmapData) [0x00000] in <filename unknown>:0 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format) [0x00000] in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) System.Drawing.Bitmap:LockBits (System.Drawing.Rectangle,System.Drawing.Imaging.ImageLockMode,System.Drawing.Imaging.PixelFormat)
  at OpenRA.Graphics.Util.FastCopyIntoSprite (OpenRA.Graphics.Sprite dest, System.Drawing.Bitmap src) [0x00000] in <filename unknown>:0 
  at OpenRA.Graphics.SheetBuilder.Add (System.Drawing.Bitmap src) [0x00000] in <filename unknown>:0 
  at OpenRA.MapCache.LoadAsyncInternal () [0x00000] in <filename unknown>:0 
  at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentException: A null reference or invalid value was found [GDI+ status: InvalidParameter]
  at System.Drawing.GDIPlus.CheckStatus (Status status) [0x00000] in <filename unknown>:0 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format, System.Drawing.Imaging.BitmapData bitmapData) [0x00000] in <filename unknown>:0 
  at System.Drawing.Bitmap.LockBits (Rectangle rect, ImageLockMode flags, PixelFormat format) [0x00000] in <filename unknown>:0 
  at (wrapper remoting-invoke-with-check) System.Drawing.Bitmap:LockBits (System.Drawing.Rectangle,System.Drawing.Imaging.ImageLockMode,System.Drawing.Imaging.PixelFormat)
  at OpenRA.Graphics.Util.FastCopyIntoSprite (OpenRA.Graphics.Sprite dest, System.Drawing.Bitmap src) [0x00000] in <filename unknown>:0 
  at OpenRA.Graphics.SheetBuilder.Add (System.Drawing.Bitmap src) [0x00000] in <filename unknown>:0 
  at OpenRA.MapCache.LoadAsyncInternal () [0x00000] in <filename unknown>:0 
  at System.Threading.Thread.StartInternal () [0x00000] in <filename unknown>:0 
</code></pre>

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/10868670-random-crash-at-startup?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

--- a/OpenRA.Game/Exts.cs
+++ b/OpenRA.Game/Exts.cs
@@ -11,6 +11,7 @@
 using System;
 using System.Collections.Generic;
 using System.Drawing;
+using System.Drawing.Imaging;
 using System.Globalization;
 using System.Linq;
 using System.Reflection;
@@ -433,6 +434,25 @@ namespace OpenRA
 		}
 
 		public static Rectangle Bounds(this Bitmap b) { return new Rectangle(0, 0, b.Width, b.Height); }
+
+		public static Bitmap CloneWith32bbpArgbPixelFormat(this Bitmap original)
+		{
+			// Note: We would use original.Clone(original.Bounds(), PixelFormat.Format32bppArgb)
+			// but this doesn't work on mono.
+			var clone = new Bitmap(original.Width, original.Height, PixelFormat.Format32bppArgb);
+			try
+			{
+				using (var g = System.Drawing.Graphics.FromImage(clone))
+					g.DrawImage(original, original.Bounds());
+			}
+			catch (Exception)
+			{
+				clone.Dispose();
+				throw;
+			}
+
+			return clone;
+		}
 
 		public static int ToBits(this IEnumerable<bool> bits)
 		{

--- a/OpenRA.Game/Graphics/Util.cs
+++ b/OpenRA.Game/Graphics/Util.cs
@@ -64,6 +64,9 @@ namespace OpenRA.Graphics
 
 		public static void FastCopyIntoSprite(Sprite dest, Bitmap src)
 		{
+			if (src.PixelFormat != PixelFormat.Format32bppArgb)
+				throw new ArgumentException("src must have a PixelFormat of Format32bppArgb", "src");
+
 			var data = dest.Sheet.GetData();
 			var dataStride = dest.Sheet.Size.Width * 4;
 			var x = dest.Bounds.Left * 4;
@@ -72,7 +75,7 @@ namespace OpenRA.Graphics
 			var height = dest.Bounds.Height;
 
 			var bd = src.LockBits(src.Bounds(),
-				ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);
+				ImageLockMode.ReadWrite, src.PixelFormat);
 			for (var row = 0; row < height; row++)
 				Marshal.Copy(IntPtr.Add(bd.Scan0, row * bd.Stride), data, (y + row) * dataStride + x, width);
 			src.UnlockBits(bd);

--- a/OpenRA.Game/Map/Map.cs
+++ b/OpenRA.Game/Map/Map.cs
@@ -398,7 +398,7 @@ namespace OpenRA
 				if (CustomPreview.PixelFormat != PixelFormat.Format32bppArgb)
 				{
 					var original = CustomPreview;
-					CustomPreview = original.Clone(original.Bounds(), PixelFormat.Format32bppArgb);
+					CustomPreview = original.CloneWith32bbpArgbPixelFormat();
 					original.Dispose();
 				}
 			}

--- a/OpenRA.Game/Map/Map.cs
+++ b/OpenRA.Game/Map/Map.cs
@@ -12,6 +12,7 @@ using System;
 using System.Collections;
 using System.Collections.Generic;
 using System.Drawing;
+using System.Drawing.Imaging;
 using System.IO;
 using System.Linq;
 using System.Security.Cryptography;
@@ -391,8 +392,16 @@ namespace OpenRA
 			DefaultSubCell = (SubCell)Game.ModData.Manifest.SubCellDefaultIndex;
 
 			if (Container.Exists("map.png"))
+			{
 				using (var dataStream = Container.GetContent("map.png"))
 					CustomPreview = new Bitmap(dataStream);
+				if (CustomPreview.PixelFormat != PixelFormat.Format32bppArgb)
+				{
+					var original = CustomPreview;
+					CustomPreview = original.Clone(original.Bounds(), PixelFormat.Format32bppArgb);
+					original.Dispose();
+				}
+			}
 
 			PostInit();
 

--- a/OpenRA.Game/Map/MapCache.cs
+++ b/OpenRA.Game/Map/MapCache.cs
@@ -11,6 +11,7 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
+using System.Drawing.Imaging;
 using System.IO;
 using System.Linq;
 using System.Net;
@@ -173,6 +174,11 @@ namespace OpenRA
 					{
 						createdPreview = true;
 						bitmap = Minimap.RenderMapPreview(modData.DefaultRules.TileSets[p.Map.Tileset], p.Map, modData.DefaultRules, true);
+					}
+					else if (bitmap.PixelFormat != PixelFormat.Format32bppArgb)
+					{
+						Log.Write("debug", "Discarding invalid map preview for {0}", p.Uid);
+						continue;
 					}
 
 					// Note: this is not generally thread-safe, but it works here because:

--- a/OpenRA.Game/Map/MapPreview.cs
+++ b/OpenRA.Game/Map/MapPreview.cs
@@ -159,7 +159,7 @@ namespace OpenRA
 						if (CustomPreview.PixelFormat != PixelFormat.Format32bppArgb)
 						{
 							var original = CustomPreview;
-							CustomPreview = original.Clone(original.Bounds(), PixelFormat.Format32bppArgb);
+							CustomPreview = original.CloneWith32bbpArgbPixelFormat();
 							original.Dispose();
 						}
 					}

--- a/OpenRA.Game/Map/MapPreview.cs
+++ b/OpenRA.Game/Map/MapPreview.cs
@@ -13,6 +13,7 @@ using System.Collections.Generic;
 using System.ComponentModel;
 using System.Diagnostics.CodeAnalysis;
 using System.Drawing;
+using System.Drawing.Imaging;
 using System.IO;
 using System.Linq;
 using System.Net;
@@ -155,6 +156,12 @@ namespace OpenRA
 						SpawnPoints = spawns;
 
 						CustomPreview = new Bitmap(new MemoryStream(Convert.FromBase64String(r.minimap)));
+						if (CustomPreview.PixelFormat != PixelFormat.Format32bppArgb)
+						{
+							var original = CustomPreview;
+							CustomPreview = original.Clone(original.Bounds(), PixelFormat.Format32bppArgb);
+							original.Dispose();
+						}
 					}
 					catch (Exception) { }
 

